### PR TITLE
Further optimise `zio.http.Body.FileBody.asStream` code

### PR DESCRIPTION
**Pull request by guizmaii**
Following https://github.com/zio/zio-http/pull/3215

### CODE REVIEW

File: zio-http/shared/src/main/scala/zio/http/Body.scala [hunk 1]
```diff
@@ -568,27 +570,32 @@ object Body {
...
+          ZIO.suspendSucceed {
+            try {
               val fs   = new FileInputStream(file)
               val size = Math.min(chunkSize.toLong, file.length()).toInt
 
-              (fs, size)
-            }
-            (fs, size) = r
-          } yield ZStream
-            .repeatZIOOption[Any, Throwable, Chunk[Byte]] {
```
Comment (guizmaii) [inline zio-http/shared/src/main/scala/zio/http/Body.scala:580]: Replacing `ZStream.repeatZIOOption` with `ZStream.repeatZIOChunkOption` allows us to remove the `.flattenChunks`

Anyway, `repeatZIOOption` is using `repeatZIOChunkOption` underneath:

Screenshot 2024-12-06 at 12 19 16 AM

File: zio-http/shared/src/main/scala/zio/http/Body.scala [hunk 2]
```diff
@@ -568,27 +570,32 @@ object Body {
...
-              for {
-                buffer <- ZIO.succeed(new Array[Byte](size))
-                len    <- ZIO.attempt(fs.read(buffer)).mapError(Some(_))
-                bytes  <-
-                  if (len > 0) ZIO.succeed(Chunk.fromArray(buffer.slice(0, len)))
-                  else ZIO.fail(None)
-              } yield bytes
+              Exit.succeed {
+                ZStream.repeatZIOChunkOption {
+                  ZIO.suspendSucceed {
```
Comment (guizmaii) [inline zio-http/shared/src/main/scala/zio/http/Body.scala]: @kyri-petrou Are we sure this `ZIO.suspendSucceed` and the `ZIO.attempt(fs.close())` will be run on a blocking thread? 🤔

File: zio-http/shared/src/main/scala/zio/http/Body.scala [hunk 3]
```diff
@@ -568,27 +570,35 @@ object Body {
...
+                    else Exit.none
+                  } catch {
+                    case e: Throwable => Exit.fail(e)
+                  }
+                }
+
+              Exit.succeed {
+                // Optimised for our needs version of `ZIO.repeatZIOChunkOption`
+                ZStream
+                  .unfoldChunkZIO(read)(_.map(_.map(_ -> read)))
```
Comment (guizmaii) [inline zio-http/shared/src/main/scala/zio/http/Body.scala:593]: @kyri-petrou Funnily enough, if we were managing to find a way to write a recursive type in Scala, 

1. we wouldn't need this ugly `(_.map(_.map(_ -> read))`
2. we could write this:
```scala
              type Read = Task[Option[(Chunk[Byte], Read)]]
              val read: Read =
                ZIO.suspendSucceed {
                  try {
                    val buffer = new Array[Byte](size)
                    val len    = fs.read(buffer)
                    if (len > 0) Exit.succeed(Some(Chunk.fromArray(buffer.slice(0, len) -> read)))
                    else Exit.none
                  } catch {
                    case e: Throwable => Exit.fail(e)
                  }
                }

              Exit.succeed {
                ZStream
                  .unfoldChunkZIO(read)(ZIO.identityFn)
                  .ensuring(ZIO.attempt(fs.close()).ignoreLogged)
              }
```

😍

Comment (987Nabil) [reply zio-http/shared/src/main/scala/zio/http/Body.scala:593]: I can imagine, that there is a nice way to do this for Scala 3. But I think we are unable to find a nice way for Scala 2

**Review by 987Nabil — APPROVED** [review]

**Pull request merged**